Schema type checker

abigailalice opened the pull request:
This should close #274 and #186, and potentially #291. The semantics are primarily based on @KaneTW's gist in #274, except with two improvements:

1. It handles array types
2. It returns an error if two columns in a `TableSchema` will have identical names, instead of passing silently

It also displays information in a slightly different way, trying to show as many errors as it can rather than bailing on the first. It also adds a number of test cases to ensure it behaves correctly for the primitive types, higher kinded types, and a few tests to ensure it fails in cases where it should fail. I'm mostly putting this forward to see what other people think, as I'm not sure if it should have other changes before being accepted. Some of my (admittedly bikesheddy) thoughts on the current state:

1. I wasn't able to re-export it from `Rel8` without creating an import cycle, so it's only exported via its own module being public. For some reason the module wouldn't compile unless it imported `Rel8`, getting a type error that `Columns` wasn't injective otherwise. I'm assuming the issue is that something not being in scope is effecting type checking, but I've never actually seen an error like that, so really aren't sure. I'm also not really sure if `Rel8.Table.Verify` is the best name for the module, I just picked `Table` pretty arbitrarily.
2. I reused `tests/Rel8/Generic/Rel8able/Test.hs` quite a bit to add some things necessary for my own test cases, but I'm not really sure if that's an appropriate place for them. I could move them if that's preferred.
3. The current work won't show KaneTW's original work, since I just copy-pasted it. Not sure if there's a proper way to base work off a gist, or if it's even an issue, but I thought it should still be said
4. I kind of want to have a version of the type checking function which also reads the data from the table after checking the schema, as having a valid schema isn't enough to ensure existing data can be read by rel8, which can be important when doing migrations. However, this isn't possible just using the `Statement` type. I just add a version using `Session` which does this, but for people using `hasql-transaction` they would still need to rewrite that function manually. This isn't a terribly difficult thing for users of the library to do themselves, since anyone using it is already going to need to convert from `Statement` to `Transaction`, and can just read from the table after type checking, but it just feels like an oversite to not have this. For the moment the issue is explained in the documentation for the function.

KaneTW commented:
There's a way to tag people in the commit text:


> Co-authored-by: NAME <NAME@EXAMPLE.COM>

For me it would be:

> Co-authored-by: David Kraeutmann <kane@kane.cx>

abigailalice commented:
Oh, I think `Fixed n` and `Char` also have a bug, due to not handling modifiers correctly. I'll work on that, as well as add the co-author disclaimer.

The failed build seems to just be from an unrelated test which happens to fail non-determinstically, using too precise rational comparison. I could fix that too, though I'm not sure if the 1e-15 is supposed to be correct, or just picked arbitrarily. I've noticed a few non-deterministic tests while working on this.

abigailalice commented:
Okay, I fixed the bugs for those corner cases, added some test cases to cover them, and added proper attribution. My only major concern now is that I needed to force push to correct the first commit message, and that may have affected other commits in the middle. I'm newish to git in groups, so let me know if that's an issue, and I can just make another branch without the issue, and issue a new pull request. Otherwise, unless any of the bikeshedding is cause for concern, I'm happy with the current state of things.

I should say I've only tested this superficially on a live project; it was able to detect some bugs in my code in some rarely inserted into tables (due to fields which had been removed in Haskell, but not in the database). But otherwise most of the testing is from the test cases themselves.

ocharles commented:
Thanks for all your hard work here @abigailalice and @KaneTW! There's quite a lot here so it might take me a while to get through this as I'm really busy at work at the moment, but I will try and get to this :pray:

abigailalice commented:
Take your time, just remember that the middle commits aren't anything related to the actual PR, only the first and last are.

ocharles commented:
> Just remember that the middle commits aren't anything related to the actual PR, only the first and last are.

This makes it very difficult to review. Could you rebase your branch so the changes are on master? Don't worry too much about commit attribution - we squash merge here anyway, so we can sort that separately.

abigailalice commented:
Okay @ocharles, I think it should be at an acceptable state to review. Sorry about the difficulty.

abigailalice commented:
@ocharles, is there something preventing this from being accepted, or have you just not had time to get around to it? Not an issue if it's the latter, just wanted to make sure if I misunderstood anything.

ocharles commented:
@abigailalice purely time constraints I'm afraid! It's still on my radar. I'll be on holiday in a few days so this might need to wait until mid September

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- table. This does not show relationships like primary or foreign keys, but can
+-- still be useful to see what types rel8 will expect of the underlying database.
+--
+-- In the event multiple columns have the same name, this will return a map of
```
```suggestion
-- In the event that multiple columns have the same name, this will return a map of
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- names to the labels identifying the column.
+checkedShowCreateTable :: Rel8able k => TableSchema (k Name) -> Either (M.Map String (NonEmpty [String])) String
+checkedShowCreateTable schema = case checkedSchemaToTypeMap schema.columns of
+    Left e -> Left $ (fmap . fmap) (\typ -> typ.label) e
```
```suggestion
    Left e -> Left $ map (\(name, typ) -> (name, typ.label)) e
```
Slightly clearer

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+equalName :: QualifiedName -> QualifiedName -> Bool
+equalName (QualifiedName a (Just b)) (QualifiedName a' (Just b'))
+  = L.dropWhile (=='_') a == L.dropWhile (=='_') a' && b == b'
+equalName (QualifiedName a _) (QualifiedName a' _)
```
```suggestion
equalName (QualifiedName a Nothing) (QualifiedName a' Nothing)
```
this is just a style nitpick

TeofilC replied:
Nope that's wrong oops. I'll drop this

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+    , duplicates :: M.Map String (NonEmpty.NonEmpty TypeInfo)
+    }
+  deriving (Show)
+data ColumnInfo = ColumnInfo
```
```suggestion

data ColumnInfo = ColumnInfo
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+    , dbType :: Maybe TypeInfo
+    , error :: Maybe ColumnError
+    } deriving (Show)
+data ColumnError
```
```suggestion

data ColumnError
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+        :: (ToExprs (k Expr) (GFromExprs k), Rel8able k)
+        => TableSchema (k Name) -> SomeTableSchema
+
+-- |@'getSchemaErrors'@ checks whether the provided schemas have correct postgresql
```
```suggestion
-- |@'getSchemaErrors'@ checks whether the provided schemas have the correct PostgreSQL
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+
+-- |@'getSchemaErrors'@ checks whether the provided schemas have correct postgresql
+-- column names and types to allow reading and writing from their equivalent Haskell
+-- types, returning a list of errors if that is not the case. The function does not
```
```suggestion
-- types, and returns a list of errors if that is not the case. The function does not
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- columns are non-null depending on the value of the tag. However, if the
+-- schema is valid rel8 shouldn't be able to write invalid data to the table.
+--
+-- However, it is possible for migrations to cause valid data to become invalid,
```
```suggestion
-- However, it is possible for migrations to cause valid data to become invalid
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+--
+-- However, it is possible for migrations to cause valid data to become invalid,
+-- in ways not detectable by this function, if the migration code changes the
+-- schema correctly but doesn't handle the value-level constraints correctly, so
```
```suggestion
-- schema correctly but doesn't handle the value-level constraints correctly. So
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- in ways not detectable by this function, if the migration code changes the
+-- schema correctly but doesn't handle the value-level constraints correctly, so
+-- it is a good idea to both read from the tables and check the schema for errors
+-- in a transaction during the migration. This former will catch value-level
```
```suggestion
-- in a transaction during the migration. The former will catch value-level
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- tables = [SomeTableSchema userTable, SomeTable orderTable]
+-- @
+--
+-- This is used by @'schemaErrors'@ to more conveniently group every table an
+-- application relies on together, for typechecking the postgresql schemas
+-- together in a single batch.
```
```suggestion
-- This is used by @'schemaErrors'@ to conveniently group every table an
-- application relies on for typechecking the PostgreSQL schema
-- together in a single batch.
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- |@'getSchemaErrors'@ checks whether the provided schemas have correct postgresql
+-- column names and types to allow reading and writing from their equivalent Haskell
+-- types, returning a list of errors if that is not the case. The function does not
+-- crash on an encountered bug, instead leaving it to the caller to decide how
```
```suggestion
-- crash on encountering a bug, instead leaving it to the caller to decide how
```

TeofilC commented on src/Rel8/Table/Verify.hs:
```diff
@@ -0,0 +1,642 @@
+-- This function does nothing to check that the conflict target of an @Upsert@
+-- are valid for the schema, nor can it prevent invalid uses of @unsafeDefault@.
+-- However, it should be enough to catch the most likely errors.
+getSchemaErrors :: [SomeTableSchema] -> HS.Statement () (Maybe Text)
```
Do we want Text or String here. Some of the other functions return String but this returns Text. I don't think it makes much of a difference but I wanted to check whether you wanted it to be consistent.
